File: tests/clients/python/driver.py
#!/usr/bin/env python3
"""Subprocess client driver for the tingly-box protocol matrix harness.

Speaks the JSON-over-stdin/stdout contract defined in
internal/protocoltest/client_subprocess.go: reads one request object from
stdin, sends it through the gateway using the REAL Python SDKs (anthropic,
openai), and writes one normalized response object to stdout.

The point of this driver is strict client-side validation: every response is
parsed by the SDKs' pydantic models. There is no raw-JSON fallback — if the
gateway emits something the real Python SDK rejects, that is the finding.

Gateway/API errors are reported in-band via the "error" field; a non-zero
exit code means the driver itself is broken.
"""

import json
import sys


def fail(msg):
    print(msg, file=sys.stderr)
    sys.exit(1)


def emit(resp):
    json.dump(resp, sys.stdout)
    sys.stdout.write("\n")


def tool_call(id_, name, arguments):
    return {"id": id_, "name": name, "arguments": arguments}


# ── Anthropic (v1 + beta) ────────────────────────────────────────────────────


def normalize_anthropic_message(msg):
    content = ""
    thinking = ""
    tool_calls = []
    for block in msg.content:
        btype = getattr(block, "type", None)
        if btype == "text":
            content += block.text
        elif btype == "thinking":
            thinking += block.thinking
        elif btype == "tool_use":
            tool_calls.append(
                tool_call(block.id, block.name, json.dumps(block.input or {}))
            )
    usage = None
    if getattr(msg, "usage", None) is not None:
        usage = {
            "input_tokens": msg.usage.input_tokens or 0,
            "output_tokens": msg.usage.output_tokens or 0,
        }
    return {
        "http_status": 200,
        "role": msg.role,
        "content": content,
        "model": str(msg.model),
        "finish_reason": msg.stop_reason or "",
        "thinking": thinking,
        "tool_calls": tool_calls,
        "usage": usage,
        "raw_body": msg.model_dump_json(),
    }


def run_anthropic(req, beta):
    import anthropic

    client = anthropic.Anthropic(
        base_url=req["base_url"] + "/tingly/anthropic",
        api_key=req["api_key"],
        max_retries=0,
        timeout=req.get("timeout_ms", 30000) / 1000.0,
    )
    svc = client.beta.messages if beta else client.messages
    kwargs = dict(
        model=req["model"],
        max_tokens=1024,
        messages=[{"role": "user", "content": req["prompt"]}],
    )
    if beta:
        # The gateway selects the Beta protocol via the ?beta=true query param.
        kwargs["extra_query"] = {"beta": "true"}

    try:
        if not req["stream"]:
            msg = svc.create(**kwargs)
            return normalize_anthropic_message(msg)

        count = 0
        try:
            with svc.stream(**kwargs) as stream:
                for _event in stream:
                    count += 1
                msg = stream.get_final_message()
            resp = normalize_anthropic_message(msg)
            resp["stream_event_count"] = count
            return resp
        except anthropic.APIStatusError:
            raise
        except Exception as e:
            # Mid-stream error event (truncated upstream): the SDK raised; the
            # turn was not completed. Report it in-band rather than crashing.
            return {"http_status": 200, "stream_event_count": count,
                    "raw_body": str(e),
                    "error": {"status": 0, "type": type(e).__name__, "message": str(e)}}
    except anthropic.APIStatusError as e:
        return api_error_response(e)


# ── OpenAI Chat Completions ──────────────────────────────────────────────────


def normalize_chat_completion(resp):
    choice = resp.choices[0]
    msg = choice.message
    tool_calls = []
    for tc in msg.tool_calls or []:
        tool_calls.append(tool_call(tc.id, tc.function.name, tc.function.arguments))
    usage = None
    if resp.usage is not None:
        usage = {
            "input_tokens": resp.usage.prompt_tokens or 0,
            "output_tokens": resp.usage.completion_tokens or 0,
        }
    return {
        "http_status": 200,
        "role": msg.role,
        "content": msg.content or "",
        "model": resp.model,
        "finish_reason": choice.finish_reason or "",
        "thinking": "",
        "tool_calls": tool_calls,
        "usage": usage,
        "raw_body": resp.model_dump_json(),
    }


def run_openai_chat(req, client):
    import openai

    kwargs = dict(
        model=req["model"],
        messages=[{"role": "user", "content": req["prompt"]}],
    )
    try:
        if not req["stream"]:
            return normalize_chat_completion(client.chat.completions.create(**kwargs))

        count = 0
        role = ""
        content = ""
        finish = ""
        model = ""
        # tool calls accumulate by index, like the Go ChatCompletionAccumulator
        tools = {}
        for chunk in client.chat.completions.create(stream=True, **kwargs):
            count += 1
            model = chunk.model or model
            if not chunk.choices:
                continue
            choice = chunk.choices[0]
            delta = choice.delta
            if delta is not None:
                role = delta.role or role
                content += delta.content or ""
                for tc in delta.tool_calls or []:
                    slot = tools.setdefault(tc.index, {"id": "", "name": "", "arguments": ""})
                    if tc.id:
                        slot["id"] = tc.id
                    if tc.function is not None:
                        if tc.function.name:
                            slot["name"] = tc.function.name
                        slot["arguments"] += tc.function.arguments or ""
            finish = choice.finish_reason or finish
        tool_calls = [
            tool_call(t["id"], t["name"], t["arguments"])
            for _, t in sorted(tools.items())
        ]
        return {
            "http_status": 200,
            "role": role,
            "content": content,
            "model": model,
            "finish_reason": finish,
            "thinking": "",
            "tool_calls": tool_calls,
            "usage": None,
            "stream_event_count": count,
            "raw_body": content,
        }
    except openai.APIStatusError as e:
        return api_error_response(e)


# ── OpenAI Responses ─────────────────────────────────────────────────────────


def normalize_response(resp):
    content = ""
    tool_calls = []
    for item in resp.output or []:
        itype = getattr(item, "type", None)
        if itype == "message":
            for part in item.content or []:
                if getattr(part, "type", None) == "output_text":
                    content += part.text
        elif itype == "function_call":
            tool_calls.append(
                tool_call(item.call_id or item.id or "", item.name, item.arguments or "")
            )
    usage = None
    if getattr(resp, "usage", None) is not None:
        usage = {
            "input_tokens": resp.usage.input_tokens or 0,
            "output_tokens": resp.usage.output_tokens or 0,
        }
    return {
        "http_status": 200,
        "role": "assistant",
        "content": content,
        "model": str(resp.model),
        "finish_reason": resp.status or "",
        "thinking": "",
        "tool_calls": tool_calls,
        "usage": usage,
        "raw_body": resp.model_dump_json(),
    }


def run_openai_responses(req, client):
    import openai

    kwargs = dict(model=req["model"], input=req["prompt"])
    try:
        if not req["stream"]:
            return normalize_response(client.responses.create(**kwargs))

        count = 0
        final = None
        delta_text = ""
        for event in client.responses.create(stream=True, **kwargs):
            count += 1
            etype = getattr(event, "type", "")
            if etype in ("response.completed", "response.incomplete"):
                final = event.response
            elif etype == "response.output_text.delta":
                delta_text += event.delta or ""
        if final is not None:
            resp = normalize_response(final)
        else:
            # Stream was cut before a terminal event (midstream-close):
            # report what was accumulated from deltas.
            resp = {
                "http_status": 200,
                "role": "assistant",
                "content": delta_text,
                "model": "",
                "finish_reason": "",
                "thinking": "",
                "tool_calls": [],
                "usage": None,
                "raw_body": delta_text,
            }
        resp["stream_event_count"] = count
        return resp
    except openai.APIStatusError as e:
        return api_error_response(e)


# ── shared ───────────────────────────────────────────────────────────────────


def api_error_response(e):
    """Map an SDK APIStatusError to the in-band error envelope."""
    try:
        raw = e.response.text
    except Exception:
        raw = str(e)
    return {
        "http_status": e.status_code,
        "raw_body": raw,
        "error": {
            "status": e.status_code,
            "type": type(e).__name__,
            "message": str(e),
        },
    }


def main():
    req = json.load(sys.stdin)
    if req.get("version") != 1:
        fail(f"unsupported driver protocol version: {req.get('version')}")

    source = req["source"]
    if source == "anthropic_v1":
        emit(run_anthropic(req, beta=False))
    elif source == "anthropic_beta":
        emit(run_anthropic(req, beta=True))
    elif source in ("openai_chat", "openai_responses"):
        import openai

        client = openai.OpenAI(
            base_url=req["base_url"] + "/tingly/openai/v1",
            api_key=req["api_key"],
            max_retries=0,
            timeout=req.get("timeout_ms", 30000) / 1000.0,
        )
        if source == "openai_chat":
            emit(run_openai_chat(req, client))
        else:
            emit(run_openai_responses(req, client))
    else:
        fail(f"unsupported source protocol: {source}")


if __name__ == "__main__":
    main()

File: tests/clients/node/driver.mjs
#!/usr/bin/env node
// Subprocess client driver for the tingly-box protocol matrix harness.
//
// Speaks the JSON-over-stdin/stdout contract defined in
// internal/protocoltest/client_subprocess.go: reads one request object from
// stdin, sends it through the gateway using the REAL Node SDKs
// (@anthropic-ai/sdk, openai), and writes one normalized response object to
// stdout.
//
// Gateway/API errors are reported in-band via the "error" field; a non-zero
// exit code means the driver itself is broken.

import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";

function toolCall(id, name, args) {
  return { id, name, arguments: args };
}

// ── Anthropic (v1 + beta) ────────────────────────────────────────────────────

function normalizeAnthropicMessage(msg) {
  let content = "";
  let thinking = "";
  const toolCalls = [];
  for (const block of msg.content ?? []) {
    if (block.type === "text") content += block.text;
    else if (block.type === "thinking") thinking += block.thinking;
    else if (block.type === "tool_use")
      toolCalls.push(toolCall(block.id, block.name, JSON.stringify(block.input ?? {})));
  }
  return {
    http_status: 200,
    role: msg.role,
    content,
    model: msg.model ?? "",
    finish_reason: msg.stop_reason ?? "",
    thinking,
    tool_calls: toolCalls,
    usage: msg.usage
      ? { input_tokens: msg.usage.input_tokens ?? 0, output_tokens: msg.usage.output_tokens ?? 0 }
      : null,
    raw_body: JSON.stringify(msg),
  };
}

async function runAnthropic(req, beta) {
  const client = new Anthropic({
    baseURL: req.base_url + "/tingly/anthropic",
    apiKey: req.api_key,
    maxRetries: 0,
    timeout: req.timeout_ms ?? 30000,
  });
  const svc = beta ? client.beta.messages : client.messages;
  const params = {
    model: req.model,
    max_tokens: 1024,
    messages: [{ role: "user", content: req.prompt }],
  };
  // The gateway selects the Beta protocol via the ?beta=true query param.
  const opts = beta ? { query: { beta: "true" } } : {};

  if (!req.stream) {
    return normalizeAnthropicMessage(await svc.create(params, opts));
  }

  const stream = svc.stream(params, opts);
  let count = 0;
  try {
    for await (const _event of stream) count++;
    const msg = await stream.finalMessage();
    const resp = normalizeAnthropicMessage(msg);
    resp.stream_event_count = count;
    return resp;
  } catch (e) {
    if (isAPIError(e)) return apiErrorResponse(e);
    // Mid-stream error event (e.g. a truncated upstream): the SDK raised; the
    // turn was not completed. Report it in-band rather than crashing.
    return { http_status: 200, stream_event_count: count, raw_body: String(e?.message ?? e),
      error: { status: 0, type: e?.name ?? "StreamError", message: String(e?.message ?? e) } };
  }
}

// ── OpenAI Chat Completions ──────────────────────────────────────────────────

function normalizeChatCompletion(resp) {
  const choice = resp.choices[0];
  const msg = choice.message;
  const toolCalls = (msg.tool_calls ?? []).map((tc) =>
    toolCall(tc.id, tc.function.name, tc.function.arguments),
  );
  return {
    http_status: 200,
    role: msg.role,
    content: msg.content ?? "",
    model: resp.model ?? "",
    finish_reason: choice.finish_reason ?? "",
    thinking: "",
    tool_calls: toolCalls,
    usage: resp.usage
      ? { input_tokens: resp.usage.prompt_tokens ?? 0, output_tokens: resp.usage.completion_tokens ?? 0 }
      : null,
    raw_body: JSON.stringify(resp),
  };
}

async function runOpenAIChat(req, client) {
  const params = {
    model: req.model,
    messages: [{ role: "user", content: req.prompt }],
  };
  if (!req.stream) {
    return normalizeChatCompletion(await client.chat.completions.create(params));
  }

  const stream = await client.chat.completions.create({ ...params, stream: true });
  let count = 0;
  let role = "";
  let content = "";
  let finish = "";
  let model = "";
  const tools = new Map();
  for await (const chunk of stream) {
    count++;
    model = chunk.model || model;
    const choice = chunk.choices?.[0];
    if (!choice) continue;
    const delta = choice.delta ?? {};
    role = delta.role || role;
    content += delta.content ?? "";
    for (const tc of delta.tool_calls ?? []) {
      const slot = tools.get(tc.index) ?? { id: "", name: "", arguments: "" };
      if (tc.id) slot.id = tc.id;
      if (tc.function?.name) slot.name = tc.function.name;
      slot.arguments += tc.function?.arguments ?? "";
      tools.set(tc.index, slot);
    }
    finish = choice.finish_reason || finish;
  }
  const toolCalls = [...tools.entries()]
    .sort((a, b) => a[0] - b[0])
    .map(([, t]) => toolCall(t.id, t.name, t.arguments));
  return {
    http_status: 200,
    role,
    content,
    model,
    finish_reason: finish,
    thinking: "",
    tool_calls: toolCalls,
    usage: null,
    stream_event_count: count,
    raw_body: content,
  };
}

// ── OpenAI Responses ─────────────────────────────────────────────────────────

function normalizeResponse(resp) {
  let content = "";
  const toolCalls = [];
  for (const item of resp.output ?? []) {
    if (item.type === "message") {
      for (const part of item.content ?? []) {
        if (part.type === "output_text") content += part.text;
      }
    } else if (item.type === "function_call") {
      toolCalls.push(toolCall(item.call_id ?? item.id ?? "", item.name, item.arguments ?? ""));
    }
  }
  return {
    http_status: 200,
    role: "assistant",
    content,
    model: resp.model ?? "",
    finish_reason: resp.status ?? "",
    thinking: "",
    tool_calls: toolCalls,
    usage: resp.usage
      ? { input_tokens: resp.usage.input_tokens ?? 0, output_tokens: resp.usage.output_tokens ?? 0 }
      : null,
    raw_body: JSON.stringify(resp),
  };
}

async function runOpenAIResponses(req, client) {
  const params = { model: req.model, input: req.prompt };
  if (!req.stream) {
    return normalizeResponse(await client.responses.create(params));
  }

  const stream = await client.responses.create({ ...params, stream: true });
  let count = 0;
  let final = null;
  let deltaText = "";
  for await (const event of stream) {
    count++;
    if (event.type === "response.completed" || event.type === "response.incomplete") {
      final = event.response;
    } else if (event.type === "response.output_text.delta") {
      deltaText += event.delta ?? "";
    }
  }
  let resp;
  if (final) {
    resp = normalizeResponse(final);
  } else {
    // Stream cut before a terminal event (midstream-close).
    resp = {
      http_status: 200,
      role: "assistant",
      content: deltaText,
      model: "",
      finish_reason: "",
      thinking: "",
      tool_calls: [],
      usage: null,
      raw_body: deltaText,
    };
  }
  resp.stream_event_count = count;
  return resp;
}

// ── shared ───────────────────────────────────────────────────────────────────

function apiErrorResponse(e) {
  const raw = e.error !== undefined ? JSON.stringify(e.error) : String(e.message ?? e);
  return {
    http_status: e.status ?? 0,
    raw_body: raw,
    error: { status: e.status ?? 0, type: e.name ?? "APIError", message: String(e.message ?? e) },
  };
}

function isAPIError(e) {
  return typeof e?.status === "number" && e.status >= 400;
}

async function readStdin() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  return Buffer.concat(chunks).toString("utf8");
}

async function main() {
  const req = JSON.parse(await readStdin());
  if (req.version !== 1) {
    throw new Error(`unsupported driver protocol version: ${req.version}`);
  }

  let resp;
  try {
    if (req.source === "anthropic_v1") {
      resp = await runAnthropic(req, false);
    } else if (req.source === "anthropic_beta") {
      resp = await runAnthropic(req, true);
    } else if (req.source === "openai_chat" || req.source === "openai_responses") {
      const client = new OpenAI({
        baseURL: req.base_url + "/tingly/openai/v1",
        apiKey: req.api_key,
        maxRetries: 0,
        timeout: req.timeout_ms ?? 30000,
      });
      resp =
        req.source === "openai_chat"
          ? await runOpenAIChat(req, client)
          : await runOpenAIResponses(req, client);
    } else {
      throw new Error(`unsupported source protocol: ${req.source}`);
    }
  } catch (e) {
    if (!isAPIError(e)) throw e;
    resp = apiErrorResponse(e);
  }
  process.stdout.write(JSON.stringify(resp) + "\n");
}

main().catch((e) => {
  console.error(e?.stack ?? String(e));
  process.exit(1);
});
